File: demo-apps/ios/VoxaTraceDemo/VoxaTraceDemo/Sections/Recording/RecordingView.swift
import SwiftUI
import VoxaTrace

/// Recording view demonstrating SonixRecorder with Config + Factory pattern.
struct RecordingView: View {
    @StateObject private var viewModel = RecordingViewModel()

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Recording")
                .font(.headline)

            Text("Status: \(viewModel.status)")
                .font(.caption)
                .foregroundColor(.secondary)

            // Format selector
            formatSelector

            Text("Duration: \(viewModel.formattedDuration)")
                .font(.body)

            // Audio level indicator
            audioLevelIndicator

            // Buffer pool status (for Calibra integration)
            if viewModel.isRecording {
                Text("Buffer Pool: \(viewModel.bufferPoolAvailable)/4 available" +
                     (viewModel.bufferPoolWasExhausted ? " (overflow occurred)" : " (zero-alloc)"))
                    .font(.caption)
                    .foregroundColor(viewModel.bufferPoolWasExhausted ? .red : .blue)
            }

            // Control buttons
            recordingControls

            // Playback section
            if let _ = viewModel.savedFilePath, !viewModel.isRecording {
                savedFileSection
            }
        }
        .onDisappear {
            viewModel.onDisappear()
        }
    }

    private var formatSelector: some View {
        HStack {
            Text("Format:")
                .font(.caption)
            ForEach(viewModel.formats, id: \.self) { format in
                formatButton(format: format)
            }
        }
    }

    @ViewBuilder
    private func formatButton(format: String) -> some View {
        let isSelected = viewModel.selectedFormat == format
        if isSelected {
            Button(format.uppercased()) {
                viewModel.selectedFormat = format
            }
            .buttonStyle(.borderedProminent)
            .controlSize(.small)
            .disabled(viewModel.isRecording)
        } else {
            Button(format.uppercased()) {
                viewModel.selectedFormat = format
            }
            .buttonStyle(.bordered)
            .controlSize(.small)
            .disabled(viewModel.isRecording)
        }
    }

    private var audioLevelIndicator: some View {
        HStack {
            Text("Level:")
                .font(.caption)
            ProgressView(value: Double(viewModel.audioLevel), total: 1.0)
        }
    }

    private var recordingControls: some View {
        HStack(spacing: 12) {
            Button("Record") {
                viewModel.startRecording()
            }
            .disabled(viewModel.isRecording)
            .buttonStyle(.borderedProminent)

            Button("Stop & Save") {
                viewModel.stopRecording()
            }
            .disabled(!viewModel.isRecording)
            .buttonStyle(.bordered)
        }
    }

    private var savedFileSection: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let path = viewModel.savedFilePath {
                let url = URL(fileURLWithPath: path)
                if let attrs = try? FileManager.default.attributesOfItem(atPath: path),
                   let fileSize = attrs[.size] as? Int64 {
                    Text("Saved: \(url.lastPathComponent) (\(fileSize / 1024) KB)")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }

            // Playback time display
            if viewModel.playbackDurationMs > 0 {
                Text("Playback: \(viewModel.formattedPlaybackTime) / \(viewModel.formattedPlaybackDuration)")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            // Playback controls
            HStack(spacing: 12) {
                Button("Play") {
                    viewModel.playRecording()
                }
                .disabled(!viewModel.canPlay)
                .buttonStyle(.bordered)
                .controlSize(.small)

                Button("Pause") {
                    viewModel.pausePlayback()
                }
                .disabled(!viewModel.isPlaying)
                .buttonStyle(.bordered)
                .controlSize(.small)

                Button("Stop") {
                    viewModel.stopPlayback()
                }
                .disabled(viewModel.savedFilePath == nil)
                .buttonStyle(.bordered)
                .controlSize(.small)
            }
        }
    }
}

/// Backward compatibility alias.
typealias RecordingSection = RecordingView

#Preview {
    RecordingView()
        .padding()
}

File: demo-apps/android/app/src/main/kotlin/com/musicmuni/voxatrace/demo/sections/recording/view/RecordingView.kt
package com.musicmuni.voxatrace.demo.sections.recording.view

import android.content.Context
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.musicmuni.voxatrace.demo.components.OptionChip
import com.musicmuni.voxatrace.demo.sections.recording.viewmodel.RecordingViewModel
import java.io.File

/**
 * Recording View - UI for audio recording demo.
 *
 * Demonstrates:
 * - SonixRecorder Builder pattern for advanced configuration
 * - Custom sample rate, channels, and bitrate
 * - Buffer Pool status monitoring (for Calibra integration)
 * - SonixPlayer for playback after recording
 */
@Composable
fun RecordingView(viewModel: RecordingViewModel = viewModel()) {
    val context = LocalContext.current

    // Stop recording/playback when leaving the screen (matches iOS onDisappear)
    DisposableEffect(Unit) {
        onDispose {
            viewModel.stopRecording()
            viewModel.stopPlayback()
        }
    }

    // Collect state from ViewModel
    val isRecording by viewModel.isRecording.collectAsStateWithLifecycle()
    val durationMs by viewModel.durationMs.collectAsStateWithLifecycle()
    val audioLevel by viewModel.audioLevel.collectAsStateWithLifecycle()
    val savedFilePath by viewModel.savedFilePath.collectAsStateWithLifecycle()
    val status by viewModel.status.collectAsStateWithLifecycle()
    val selectedFormat by viewModel.selectedFormat.collectAsStateWithLifecycle()
    val bufferPoolAvailable by viewModel.bufferPoolAvailable.collectAsStateWithLifecycle()
    val bufferPoolWasExhausted by viewModel.bufferPoolWasExhausted.collectAsStateWithLifecycle()
    val isPlaying by viewModel.isPlaying.collectAsStateWithLifecycle()
    val playbackTimeMs by viewModel.playbackTimeMs.collectAsStateWithLifecycle()
    val playbackDurationMs by viewModel.playbackDurationMs.collectAsStateWithLifecycle()

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "Recording",
            style = MaterialTheme.typography.titleMedium
        )

        Text(text = "Status: $status", style = MaterialTheme.typography.bodySmall)

        // Format selection
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Format:", style = MaterialTheme.typography.bodySmall)
            viewModel.formats.forEach { format ->
                OptionChip(
                    selected = selectedFormat == format,
                    onClick = { viewModel.setSelectedFormat(format) },
                    label = format.uppercase(),
                    enabled = !isRecording
                )
            }
        }

        // Duration display
        Text(
            text = "Duration: ${RecordingViewModel.formatDuration(durationMs)}",
            style = MaterialTheme.typography.bodyMedium
        )

        // Audio level indicator
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text("Level:", style = MaterialTheme.typography.bodySmall)
            LinearProgressIndicator(
                progress = { audioLevel.coerceIn(0f, 1f) },
                modifier = Modifier
                    .weight(1f)
                    .height(8.dp),
            )
        }

        // Buffer pool status (demonstrates zero-allocation processing for Calibra integration)
        if (isRecording) {
            Text(
                text = "Buffer Pool: $bufferPoolAvailable/4 available" +
                        if (bufferPoolWasExhausted) " (overflow occurred)" else " (zero-alloc)",
                style = MaterialTheme.typography.bodySmall,
                color = if (bufferPoolWasExhausted) MaterialTheme.colorScheme.error
                else MaterialTheme.colorScheme.primary
            )
        }

        // Control buttons
        Row(
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Button(
                onClick = { viewModel.startRecording(context) },
                enabled = !isRecording
            ) {
                Text("Record")
            }

            Button(
                onClick = { viewModel.stopRecording() },
                enabled = isRecording
            ) {
                Text("Stop & Save")
            }
        }

        // Show saved file info and playback controls (match iOS: show when savedFilePath exists and not recording)
        if (savedFilePath != null && !isRecording) {
            val file = File(savedFilePath!!)
            if (file.exists()) {
                Text(
                    text = "Saved: ${file.name} (${file.length() / 1024} KB)",
                    style = MaterialTheme.typography.bodySmall
                )
            }

            // Playback time display
            if (playbackDurationMs > 0) {
                Text(
                    text = "Playback: ${RecordingViewModel.formatDuration(playbackTimeMs)} / ${RecordingViewModel.formatDuration(playbackDurationMs)}",
                    style = MaterialTheme.typography.bodySmall
                )
            }

            // Playback controls
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Button(
                    onClick = { viewModel.playRecording() },
                    enabled = !isRecording && !isPlaying && file.exists()
                ) {
                    Text("Play")
                }

                Button(
                    onClick = { viewModel.pausePlayback() },
                    enabled = isPlaying
                ) {
                    Text("Pause")
                }

                Button(
                    onClick = { viewModel.stopPlayback() },
                    enabled = playbackDurationMs > 0
                ) {
                    Text("Stop")
                }
            }
        }
    }
}
